Add ability to select a crop area

serhii-londar opened the pull request:
Closes #6

sindresorhus commented:
Thanks for working on this 🙌

sindresorhus commented:
The controls disappear for a second when you shake to get a random image. This doesn't happen in `master`.

serhii-londar commented:
> The controls disappear for a second when you shake to get a random image. This doesn't happen in `master`.

Thank for your reply! I will check it.

serhii-londar commented on Blear/util.swift:
```diff
@@ -250,3 +262,18 @@ extension UIViewController {
 }
+
+extension UIScrollView {
+	func snapshot() -> UIImage? {
```
Added extension to UIScrollView as there is additional logic for visible bounds calculation.

serhii-londar commented on Blear/util.swift:
```diff
@@ -24,6 +24,19 @@ func delay(seconds: TimeInterval, closure: @escaping () -> Void) {
 	DispatchQueue.main.asyncAfter(deadline: .now() + seconds, execute: closure)
 }
 
+func calculateSize(with imageSize: CGSize, and screenSize: CGSize) -> CGSize {
```
Add `CGSize` extension with method named `func aspectFitSize(to otherSize: CGSize) -> CGSize`. Is it better now?

serhii-londar commented:
@sindresorhus Issue with controls disappearing fixed in the last commit. Also added changes related to your comments.

sindresorhus commented:
I think there needs to be some kind of indication to the user that they can scroll.

sindresorhus commented:
I just tried this on my iPhone XS and the app crashes a lot now (didn't previously) when using the slider quickly.

serhii-londar commented on Blear/util.swift:
```diff
@@ -265,16 +253,26 @@ extension UIViewController {
+			height = width / aspectRatio
+		}
+		return CGSize(width: width, height: height)
 	}
```
Updated 👍

serhii-londar commented:
> I just tried this on my iPhone XS and the app crashes a lot now (didn't previously) when using the slider quickly.

It crashed due to a memory issue. While we moving slider we process image a lot of times. That cause crash.

serhii-londar commented:
> @sindresorhus Issue with controls disappearing fixed in the last commit. Also added changes related to your comments.

How about enabling scroll view indicators?

serhii-londar commented:
> I think there needs to be some kind of indication to the user that they can scroll (not modal alert).

@sindresorhus How about enabling scroll view indicators?

sindresorhus commented:
Can we flash the scroll indicators briefly?

serhii-londar commented:
> Can we flash the scroll indicators briefly?

I thought about this. It will looks like this:

sindresorhus commented:
Hmm, that's not very visible. How about we just scroll the image with animation from left to right the first time the user uses the app?

sindresorhus commented:
https://github.com/sindresorhus/blear/pull/8/commits/489a0b5ec8b72ccfa87fa3151859d53b1f74e6f9 fixed the crashes 👌

I noticed that it's still very flickery when moving the slider. Maybe we should work on a lower resolution version of the picture when previewing and only apply the effect to the real image when saving?

serhii-londar commented:
> Hmm, that's not very visible. How about we just scroll the image with animation from left to right the first time the user uses the app?

Ok, will add.

serhii-londar commented:
> I noticed that it's still very flickery when moving the slider. Maybe we should work on a lower resolution version of the picture when previewing and only apply the effect to the real image when saving?

I'll play with that.

serhii-londar commented:
@sindresorhus I've added intro for scroll view and show it after alert with a tip for shaking displayed. Also, I've changed the source image size. Works well now on my iPhone 7 Plus. Please check.

sindresorhus commented:
Thanks for working on this 🙌
